Add tokamak example, first with no currents

bendudson opened the pull request:
Starting with no-current case, just transport with neutrals.

No grid file included, so needs to be added by user.

Includes README file

Replaces #27

bendudson committed:
Add tokamak example

Starting with no-current case, just transport with neutrals.

No grid file included, so needs to be added by user.

Includes README file

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+
+TwistShift = false  # use twist-shift condition?
+ballooning = true
+shiftinitial = false
```
Are these all necessary? I'd hope `TwistShift` and `ballooning` would be correct by default. I don't remember what the default of `shiftinitial` is.

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+[mesh]
+paralleltransform = shifted
+
+symmetricGlobalX = true
```
this is the default value.

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+
+symmetricGlobalX = true
+
+shiftangle_from_zshift=true
```
If I remember right, this was a workaround for hypnotoad2 grids when they did not include `ShiftAngle`. hypnotoad2 does calculate and save `ShiftAngle` now, so I think this can go.

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+##################################################
+# derivative methods
+
+[ddx]
```
This (and `[ddy]` and `[ddz]`) shoud be `[mesh:ddx]` (etc.) if we want them to do anything.

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+
+[ddz]
+
+first = C2
+second = C2
```
Do we definitely want these to be `C2` rather than `FFT`?

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+# Time-integration solver
+
+[solver]
+mxstep = 1000000
+
```
`mxstep` is repeated below
```suggestion
```

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+RTOL = 1.0e-5   # relative tolerance
+mxstep = 1000000  # Maximum internal steps per output
+
+[laplace]  # This is used for Delp2 operator
+all_terms = true
+nonuniform=true
```
I think these should be `true` by default now?

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+# Electrostatic potential solver
+
+[phiSolver]
+inner_boundary_flags = 0
```
Shouldn't this be `INVERT_SET` too?

bendudson replied:
Not sure: In the inversion we can have one boundary zero-gradient, just not both. It may work with INVERT_SET, but for now I've been using neumann on the inner boundary.

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+inner_boundary_flags = 0
+outer_boundary_flags = 16  # INVERT_SET  (2.8 * Te)
+
+all_terms = true
+nonuniform=true   # NOTE: Necessary to avoid numerical instability
```
True by default now?

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+atol = 1e-12
+rtol = 1e-8
+
+core_bndry_dirichlet = false
```
In case at some point the input file is modified to turn currents on - if we expect to use the evolving phi boundary conditions, should this be `true`?

bendudson replied:
Currently set to zero-gradient in the core, then evolving just outer boundary. Definitely should note this somewhere...

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+inner_boundary_flags = 0
+outer_boundary_flags = 0
+
+all_terms = true
+nonuniform=true
```
True by default?

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+
+excitation = true # Hydrogen neutral excitation radiation
+
+## Settings for 2D parallel closures
+sinks = false
+sink_invlpar =  0.2  # 5m parallel connection length
+sheath_closure = false
+drift_wave = false
```
Not needed here? "2D" in the comment could be confusing when this input file is set up for 2D (axisymmetric) simulations...

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+
+#######################
+# Plasma sheath
+sheath_model = 2  # 0 = Bohn, 1 = Loizu, 2 = Bohm + free density
```
```suggestion
sheath_model = 2  # 0 = Bohm, 1 = Loizu, 2 = Bohm + free density
```

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+startprofiles = true
+
+core_sources = true # Only sources in the core
+adapt_source_p = false   # Feedback on profiles (PI controller)
+adapt_source_n = true
```
slightly more detailed comments might be nice?

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+source_i = 1e-6       # Integral of error (1/timescale^2)
+source_vary_g11 = true  # Vary source in poloidal angle to better match radial transport
+
+staggered = false
```
I can see the setting exists, but looks like it doesn't do anything? Maybe clearer for new users to just delete (here and/or `bool staggered` in the source code) so it doesn't look like something that might be worth changing?

johnomotani commented on tokamak/1-no-currents/BOUT.inp:
```diff
@@ -0,0 +1,309 @@
+Nnorm = 1e20
+Tnorm = 100
+Bnorm = 1
+AA = 1  # Atomic mass. 1 = Hydrogen, 2 = Deuterium
```
Don't we want Deuterium for TCV-like cases?

johnomotani reviewed (commented):
Nice to have an example! I've suggested removing a few options if they're set correctly by default. I think that makes things clearer for a new user reading the file.

bendudson committed:
Update tokamak/1-no-currents/BOUT.inp

Co-authored-by: johnomotani <john.omotani@ukaea.uk>

bendudson committed:
Tidying up tokamak example inputs

Many settings not used, or not documented. Thanks @johnomotani,
this is hopefully much easier to understand now.

Still includes some options which are not needed for transport
simulations. Perhaps these could be removed.

bendudson committed:
Remove unused option staggered

Not used anywhere in the code.

bendudson committed:
Simplifying no-current input file

Removing defaults which needn't be changed, and removing
switches for cases with currents.

bendudson committed:
Add input file for case with currents on

Some instructions for copying restart files, and restarting
with the new settings.

johnomotani commented on tokamak/2-currents/BOUT.inp:
```diff
@@ -0,0 +1,273 @@
+FiniteElMass    = true  # Finite electron mass?
+
+# Electrostatic potential
+split_n0 = false  # Solve n=0 electrostatic potential separately?
```
Is this still `false` because that's more stable to get the simulations started? I'd have expected the poloidal derivatives for n=0 to be important in an axisymmetric simulation?

bendudson replied:
It can be set to `true` but then needs PETSc which I don't have mostly. It can make quite a difference, mainly around the X-point. Perhaps a separate input with these options turned on would be good.

johnomotani commented on tokamak/2-currents/BOUT.inp:
```diff
@@ -0,0 +1,273 @@
+ion_viscosity = false     # Viscous terms in vorticity equation
+ion_viscosity_par = true  # Parallel diffusion of ion momentum
+thermal_conduction = true   # Parallel heat conduction
+poloidal_flows = false   # Include ExB flows in the X-Y plane?
```
If we do `split_n0 = true`, I guess this should be `true` as well?

bendudson replied:
Not sure if they need to be related. Perhaps.

bendudson committed:
Add python script to modify curvature

Seems to be needed for simulations with drifts. It would be even
better to ensure that Div(Curl(b/B)) is zero by adjusting bxcvy.
